File: src/net/dinkla/raytracer/math/MathUtils.java
package net.dinkla.raytracer.math;

/**
 * Created by IntelliJ IDEA.
 * User: jorndinkla
 * Date: 26.04.2010
 * Time: 09:00:51
 * To change this template use File | Settings | File Templates.
 */
public class MathUtils {

    public static final double INV_PI = (1.0 / Math.PI);

    public static final double PI_ON_180 = (Math.PI / 180);

    public static final double K_HUGEVALUE = 1.0E10;

    static public final double K_EPSILON = 0.01;

    public static double clamp(double x, double low, double high) {
        return (x < low) ? low : ((x > high) ? high : x);
    }

    public static Point3D minMin(Point3D p, Point3D q, Point3D r) {
        double x = Math.min(Math.min(p.getX(), q.getX()), r.getX());
        double y = Math.min(Math.min(p.getY(), q.getY()), r.getY());
        double z = Math.min(Math.min(p.getZ(), q.getZ()), r.getZ());
        return new Point3D(x, y, z);
    }

    public static Point3D maxMax(Point3D p, Point3D q, Point3D r) {
        double x = Math.max(Math.max(p.getX(), q.getX()), r.getX());
        double y = Math.max(Math.max(p.getY(), q.getY()), r.getY());
        double z = Math.max(Math.max(p.getZ(), q.getZ()), r.getZ());
        return new Point3D(x, y, z);
    }

    public static boolean isZero(double r) {
        return r > -K_EPSILON && r < K_EPSILON;
    }

}

File: src/main/kotlin/net/dinkla/raytracer/math/MathUtils.kt
package net.dinkla.raytracer.math

import kotlin.math.max

object MathUtils {

    const val INV_PI = 1.0 / Math.PI

    const val INV_TWO_PI = 1.0 / (2.0 * Math.PI)

    const val PI_ON_180 = Math.PI / 180

    const val K_HUGEVALUE = 1.0E10

    const val K_EPSILON = 0.01

    fun clamp(x: Double, low: Double, high: Double): Double = if (x < low) low else if (x > high) high else x

    fun min(p: Point3D, q: Point3D, r: Point3D): Point3D {
        val x = min(p.x, q.x, r.x)
        val y = min(p.y, q.y, r.y)
        val z = min(p.z, q.z, r.z)
        return Point3D(x, y, z)
    }

    internal fun min(x: Double, y: Double, z: Double): Double = if (x < y) {
        if (x < z) x else z
    } else {
        if (y < z) y else z
    }

    fun maxMax(p: Point3D, q: Point3D, r: Point3D): Point3D {
        val x = max(p.x, q.x, r.x)
        val y = max(p.y, q.y, r.y)
        val z = max(p.z, q.z, r.z)
        return Point3D(x, y, z)
    }

    internal fun max(x: Double, y: Double, z: Double): Double = if (x > y) {
        if (x > z) x else z
    } else {
        if (y > z) y else z
    }

    fun isZero(x: Double): Boolean = x > -K_EPSILON && x < K_EPSILON
}
